(2)‎ 어느 쌍성의 궤도면이 관측자의 수평(x)축에 대해서 기울기각 $i$만큼 기울어진 경우 관측된 상대속도와 상대변위에 의해서 쌍성의 궤도 인자들과 총질량을 구할 수 있다. 여기에는 뉴턴역학이 가정된다. 아래 주어진 식들을 주어진 속도 성분들 및 변위 성분들에 대해서 수치적으로 풀어서 인자들을 구하라. [총 50점]

$\frac{-\nu_{z}}{\nu_{y}} = tan{i}$

$\phi = arctan((\Delta{y}/\Delta{x})/cos{i})$

$\frac{\nu_{z}}{\nu_{y}} = - cos{i} \, \frac{cos{\phi}\, + \, ecos{\phi_{0}}} {sin{\phi} \, + \,esin{\phi_{0}}}$

$\Delta{z} = -\sqrt{(\Delta{x})^2\, + \, (\Delta{y})^2}\, sin{\phi} sin{i}\, / \sqrt{(cos{\phi})^2\, + \, sin{\phi})^2(cos{i})^2}$

$v = \sqrt{v_{x}^2 \, + \, v_{y}^2 \, + \, v_{z}^2} = 941.9 \, \sqrt{\frac{M}{r} \left (2 \, - \, \frac{1-e^2}{1\,+\,ecos(\phi\, - \, \phi_{0})} \right)} \, ms^{-1}$

위에서 $r\, = \, \sqrt{(\Delta{x})^2\,+\,(\Delta{y})^2\,+\, (\Delta{z})^2}$로서 kau(kilo astronomical units) 단위로 주어지고 M은 태양질량 단위로 주어지는 쌍성의 총질량이다.

주어진 값들: $(\Delta{x}\,,\Delta{y}\, ,\Delta{z}) = (10.48\, ,4.47\, ,7.91)kau$

$(v_{x}\, ,v_{y}\, ,v_{z}) = (-90.4\, ,79.5\, ,137.7)ms^{-1}$

값을 구해야 할 인자들: $i, \phi_{0}, \phi, e, M$

이심율 e의 범위는 $0 < e < 1$이다.

각들을 나타내는 $i, \phi_{0}, \phi$는 도(degree)의 단위로 변환하여 출력하라.

경사각 $i$의 범위는 도의 단위로 $-90 < i < 90$이다.

두 각들 $\phi_{0},\,\phi$는 도의 단위로 (-360, 360) 범위 내에서 구하라.

In [ ]:
import numpy as np
from scipy.optimize import fsolve

Au_to_meter = 149597870691 #UnitConverters.net 사이트 참고해서, Au를 m로 바꾸는 수를 결정함.

delta_x_meter = 10.48 * 1000 * Au_to_meter
delta_y_meter = 4.47 * 1000 * Au_to_meter
delta_z_meter = 7.91 * 1000 * Au_to_meter

v_x = -90.4
v_y = 79.5
v_z = 137.7

# 궤도 기울기 i 계산 (Radian)
i_radian = np.arctan2(-v_z , v_y) # numpy의 arctan 함수는 결과가 -90도에서 90도 사이로만 나오기 때문에, 1사분면과 4사분면에 대해서만 알 수 있지만, arctan2를 사용하면 -v_z와 v_y 부호를 둘 다 고려해서 네 개의 사분면 모든 각도를 정확하게 계산하기 때문에 궤도 각도 계산에 더욱 적합하여 이를 사용하였다.
#일단 계산들은 radian으로 진행할 거고, 마지막 출력에는 degree로 변환해서 출력할 계획이다.

# 궤도상의 각도 \phi 계산 (Radian)

phi_radian = np.arctan2(delta_y_meter / np.cos(i_radian), delta_x_meter) # numpy의 arctan 함수는 결과가 -90도에서 90도 사이로만 나오기 때문에, 1사분면과 4사분면에 대해서만 알 수 있지만, arctan2를 사용하면 delta_x_meter와 delta_y_meter 부호를 둘 다 고려해서 네 개의 사분면 모든 각도를 정확하게 계산하기 때문에 궤도 각도 계산에 더욱 적합하여 이를 사용하였다.
# 또한 각도를 구할 때, phi_radian = np.arctan2(delta_y_meter, np.cos(i_radian) * delta_x_meter)와 같이 계산하지 않았는데, 그 이유는 지금 궤도면이 경사각 i만큼 x축에서 기울어졌기 때문에, 관측된 delta_y_meter 값은 궤도면에 대한 수직 방향으로 왜곡되어 있다. 그래서, delta_y_meter를 cos i로 나눠서 보정을 해서, 궤도면 상의 실제 각도 \phi 정확히 계산할 수 있도록 하기 위함이다.
# 반면에, phi_radian = np.arctan2(delta_y_meter, np.cos(i_radian) 이렇게 쓰면 delta_x_meter를 보정하게 되어서, 기하학적으로 궤도면의 길이 방향과 맞지 않아서 각도의 계산에 오차가 생긴다.



# 뒤에 계산에 사용할 값들 미리 구해두기

# 위치 벡터 r의 크기 구하기
r_magnitude = np.sqrt(delta_x_meter**2 + delta_y_meter**2 + delta_z_meter**2)

# 속도 벡터 v의 제곱, v^2 구하기
v_magnitude_square = v_x**2 + v_y**2 + v_z**2

#천체 역학에 유용한 Specific angular momentum h 구하기. - WIKIPEDIA의 'Specific angular momentum' 문서를 참고했다. h는 단위 질량당 각운동량(specific angular momentum)으로, h = r \times v = \frac{L}{m} | (h = r(상대 위치 벡터) ⨉ v(상대 속도 벡터) = L(총 각운동량 벡터/m(질량)) | 우선 latex로 한 번 설명하고, \times, 곱하기 기호는 관계식 설명을 위해 특수 문자를 사용했습니다.
#The specific relative angular momentum is defined as the cross product of the relative position vector r and the relative velocity vector v, where L is the angular momentum vector, defined as r × mv
# 단위 질량당 각운동량 h의 벡터 성분 - r과 v의 벡터 곱
hx = delta_y_meter * v_z - delta_z_meter * v_y
hy = delta_z_meter * v_x - delta_x_meter * v_z
hz = delta_x_meter * v_y - delta_y_meter * v_x

# 단위 질량당 각운동량 h의 제곱, h^2 구하기
h_magnitude_square = hx**2 + hy**2 + hz**2



# fsolve를 사용해서 수치적으로 풀 때 필요한 방정식들을 정의하기

def astrophysical_equation(variables):
    e, phi0_radian, M_kg = variables


    # 첫 번째 방정식 : 문제에서 주어진 $\frac{v_{z}}{v_{y}} = - cos{i} \, \frac{cos{\phi}\, + \, ecos{\phi_{0}}} {sin{\phi} \, + \,esin{\phi_{0}}}$ 사용
    # fsolve 기능을 사용하기 위해, 이 식을 모두 좌변으로 넘기고 이를 g라는 함수라고 했을 때, g = 0으로 만드는 작업을 한다.
    # $\frac{v_y}{v_x}\, + \, cos{i} \, \frac{cos{\phi}\, + \, ecos{\phi_{0}}} {sin{\phi} \, + \,esin{\phi_{0}}} = 0$
    # $\fac{v_y}{v_x}/- cos{i} * {sin{\phi} \, + \,esin{\phi_{0}} - ecos{\phi_{0}}} {sin{\phi}
    Left_term = (v_y / v_x) * (-1 / np.cos(i_radian))
    Numerator = np.cos(phi_radian) + e * np.cos(phi0_radian)
    Denominator = np.sin(phi_radian) + e * np.sin(phi0_radian)
    astrophysical_equation_1 = Left_term * Denominator - Numerator


    # 두 번째 방정식과 세 번째 방정식에서는 진근점 이각(true anomaly)을 쓸 것이기 때문에 여기에 정의해둔다.
    # WIKIPEDIA의 'True anomaly' 문서의 해당 문구를 참고했다. In celestial mechanics, true anomaly is an angular parameter that defines the position of a body moving along a Keplerian orbit.
    # It is the angle between the direction of periapsis and the current position of the body, as seen from the main focus of the ellipse (the point around which the object orbits).
    # 요약해서, 진근점 이각((true anomaly)은 천체의 궤도에서 현재 위치각(\phi)와 근일점 방향각(\phi_{0})의 차이를 의미하는데, 이는 속도랑 거리 계산에 필수적으로 사용되는 천체물리 변수라서 여기에 정의해둔다.
    true_anomaly = phi_radian - phi0_radian



    # 두 번째 방정식 : 각운동량 보존 법칙 - WIKIPEDIA의 'Orbit determination' 문서를 참고했다.
    # where \mu = GM is the standard gravitational parameter for the central body of mass  M, and  G is the universal gravitational constant.
    # Compute the semi-latus rectum p of the orbit, and its semi-major axis  a (if it is not a parabolic orbit, where  e = 1 and  a is undefined or defined as infinity)

     #WIKIPEDIA의 'Kepler orbit' 문서를 참고했다.
    # More generally, the path of an object undergoing Keplerian motion may also follow a parabola or a hyperbola, which, along with ellipses, belong to a group of curves known as conic sections.
    # Mathematically, the distance between a central body and an orbiting body can be expressed as: r(\theta) = a * (1 - e^2) / 1 + e*cos(\theta) = p / 1 + e * cos(\theta)

    #여기서 p는 (semi-latus rectum, 반통경)이고, a는 장반경, e는 이심률, theta는 true anomaly (진근점 이각), r은 거리, 즉 우리가 구한 위치 벡터의 크기 r_magnitude, \mu는 \mu = GM : 중력 파라미터
    # G는 중력상수 universal gravitational constant = 6.67430e-11 (m^3 * kg^-1 * s^-2)
    # p = h^2 / \mu = a * (1 - e^2) = r * (1 + e * cos(\theta))
    # a = p / 1 - e^2
    # h^2 = G * M * p
    # 그래서 이러한 식을 이용해서 두 번째 방정식을 세우겠다.

    G = 6.67430e-11 # WIKIPEDIA의 'Gravitational constant' 문서를 참고했다.

    p = r_magnitude * (1 + e * np.cos(true_anomaly)) # p = r * (1 + e * cos(\theta)) 관계식 이용
    astrophysical_equation_2 = h_magnitude_square - G * M_kg * p # h^2 = G * M * p 관계식 이용



    # 세 번째 방정식 : Vis-viva 방정식 (v^2 = GM(2/r - 1/a)) - WIKIPEDIA의 'Vis-viva equation' 문서를 참고했다.
    # For any Keplerian orbit (elliptic, parabolic, hyperbolic, or radial), the vis-viva equation is as follows: v^2 = GM(1/r - 1/a)
    # v는 is the relative speed of the two bodies이고, 이는 위에서 미리 계산해뒀던 v_magnitude_square = v_x**2 + v_y**2 + v_z**2에 root를 취한 꼴이다.
    # 그리고 나머지 문자들의 의미는 두 번째 방정식에서 밝힌 것과 똑같다.

    astrophysical_equation_3 = v_magnitude_square * p - G * M_kg * (2 * p / r_magnitude - (1 - e**2)) # 방정식 2에서 밝힌 a = p / 1 - e^2를 1/a = (1 - e**2) / p로 변형해서, (v^2 = GM(2/r - 1/a)) 식에 대입하고 양변에 p를 곱했다.
    # 그 이유는, 이제 뒤에서 할 fsolve는 반복적으로 다양한 시도값을 넣으며 해를 찾는데, 이 과정에서 만약에 M이 0에 가까운 값으로 잠깐이라도 접근하면, 원래 형태의 Vis-viva 방정식 v^2 = GM(2/r - 1/a)의 우변이 너무 작아져서 수치적으로 불안정해진다.
    # 그래서 이것을 방지하기 위해서 양변에 p를 곱해서, v^2 * p = GM(2p/r - (1 - e^2)) 형태로 변형하게 되면, M에 곱해지는 항의 크기가 p를 곱하기 전인 (2/r - (1 - e^2)/p)보다는 커지면서 민감도가 완화되기 때문에, 수치적으로 안정성과 수렴성이 향상되기 때문에, 양변에 p를 곱했다.

    return [astrophysical_equation_1, astrophysical_equation_2, astrophysical_equation_3]



# 이제, fsolve 기능을 이용해서 해 구하기

M_sun_kg = 1.9884156e30 # 1 태양질량 (kg) - WIKIPEDIA의 'Solar mass' 문서를 참고했다.

# 초기값 설정, (e, phi0_rad, M_kg)
# 궤도 이심률 e는 원형과 타원형 한 쪽으로 치우치지 않은 중앙값인 0.5으로 초기값 설정.
# 근일점 방향각 phi0_radian을 기준 축과 일치시켜 \phi와의 상대적 위치 관계를 단순화한 상태에서 계산을 시작하기 위해 초기값을 0으로 설정.
# M은 문제에서 '태양질량 단위로 주어지는 쌍성의 총질량'이라고 줬으니까, 가장 간단하게 태양 질량에 1를 곱한 값으로 초기값을 설정.
initial_value = [0.5, 0.0, 1.0 * M_sun_kg]

# fsolve 사용해서 연립방정식의 해를 찾기!
result = fsolve(astrophysical_equation, initial_value)

#fsolve를 통해 얻은 결과 result에서 각각 궤도 이심률 e, 근일점 방향각 phi0_radian, 쌍성 총질량 M_kg을 꺼내서 각각의 변수에 할당.
e_result = result[0]
phi0_radian_result = result[1]
M_kg_result = result[2]


# 결과를 문제 조건에 맞게 단위 변환

i_degree = np.rad2deg(i_radian) # radian을 degree로 변환
phi_degree = np.rad2deg(phi_radian) # radian을 degree로 변환
phi0_degree_result = np.rad2deg(phi0_radian_result) # radian을 degree로 변환
M_solar_result = M_kg_result / M_sun_kg # 쌍성 총질량 M_kg를 M_sun_kg로 나눠서 Solar Mass로 단위 변환

# 결과로 얻은 phi, phi0의 각도가 -360도 < phi, phi0 < 360를 만족하지 않을 수 있기 때문에 각도 범위 조정하는 코드를 짜보았다. - 근데 이거 없이 코드를 돌려도, 결과는 똑같이 나왔다. 하지만 공부 차원에서, 그리고 혹시 모르니까 만들어둔다.
phi_degree = (phi_degree + 360) % 720 - 360
phi0_degree_result = (phi0_degree_result + 360) % 720 - 360

# 각도 범위를 조정하는 위의 코드 대신 아래처럼 코드를 짤까도 생각해봤는데, -360도보다 작은 값들이 있을 수도 있긴 하지만, 그래도 360를 더하고 시작하면 웬만한 음수값들은 대부분 양수가 될 것 같아서, 180이 아닌 360을 더했다.
#phi_degree = (phi_degree + 180) % 360 - 180
#phi0_degree_result = (phi0_degree_result + 180) % 360 - 180

print(f"궤도 기울기 (i) : {i_degree:.10f} degrees")
print(f"근일점 방향각 (φ₀) : {phi0_degree_result:.10f} degrees")
print(f"궤도상의 각도 (φ) : {phi_degree:.10f} degrees")
print(f"궤도 이심률 (e) : {e_result:.10f}")
print(f"쌍성 총질량 (M) : {M_solar_result:.10f} Solar Mass")

궤도 기울기 (i) : -60.0003532839 degrees
근일점 방향각 (φ₀) : 4.5199852469 degrees
궤도상의 각도 (φ) : 40.4662526670 degrees
궤도 이심률 (e) : 0.4699954451
쌍성 총질량 (M) : 0.3643278901 Solar Mass


.



.

이번 과제의 제 코드는 자세히 설명하다 보니, 각주가 많이 길어져서 조교님이 보시기에 불편하실 수도 있을 것 같아, 각주를 제거한 버전도 함께 만들어둡니다.

In [ ]:
import numpy as np
from scipy.optimize import fsolve

Au_to_meter = 149597870691

delta_x_meter = 10.48 * 1000 * Au_to_meter
delta_y_meter = 4.47 * 1000 * Au_to_meter
delta_z_meter = 7.91 * 1000 * Au_to_meter

v_x = -90.4
v_y = 79.5
v_z = 137.7


i_radian = np.arctan2(-v_z , v_y)

phi_radian = np.arctan2(delta_y_meter / np.cos(i_radian), delta_x_meter)

r_magnitude = np.sqrt(delta_x_meter**2 + delta_y_meter**2 + delta_z_meter**2)

v_magnitude_square = v_x**2 + v_y**2 + v_z**2

hx = delta_y_meter * v_z - delta_z_meter * v_y
hy = delta_z_meter * v_x - delta_x_meter * v_z
hz = delta_x_meter * v_y - delta_y_meter * v_x

h_magnitude_square = hx**2 + hy**2 + hz**2

def astrophysical_equation(variables):
    e, phi0_radian, M_kg = variables

    Left_term = (v_y / v_x) * (-1 / np.cos(i_radian))
    Numerator = np.cos(phi_radian) + e * np.cos(phi0_radian)
    Denominator = np.sin(phi_radian) + e * np.sin(phi0_radian)
    astrophysical_equation_1 = Left_term * Denominator - Numerator


    true_anomaly = phi_radian - phi0_radian


    G = 6.67430e-11

    p = r_magnitude * (1 + e * np.cos(true_anomaly))
    astrophysical_equation_2 = h_magnitude_square - G * M_kg * p


    astrophysical_equation_3 = v_magnitude_square * p - G * M_kg * (2 * p / r_magnitude - (1 - e**2))

    return [astrophysical_equation_1, astrophysical_equation_2, astrophysical_equation_3]



M_sun_kg = 1.9884156e30


initial_value = [0.5, 0.0, 1.0 * M_sun_kg]


result = fsolve(astrophysical_equation, initial_value)


e_result = result[0]
phi0_radian_result = result[1]
M_kg_result = result[2]


i_degree = np.rad2deg(i_radian)
phi_degree = np.rad2deg(phi_radian)
phi0_degree_result = np.rad2deg(phi0_radian_result)
M_solar_result = M_kg_result / M_sun_kg


phi_degree = (phi_degree + 360) % 720 - 360
phi0_degree_result = (phi0_degree_result + 360) % 720 - 360


print(f"궤도 기울기 (i) : {i_degree:.10f} degrees")
print(f"근일점 방향각 (φ₀) : {phi0_degree_result:.10f} degrees")
print(f"궤도상의 각도 (φ) : {phi_degree:.10f} degrees")
print(f"궤도 이심률 (e) : {e_result:.10f}")
print(f"쌍성 총질량 (M) : {M_solar_result:.10f} Solar Mass")

궤도 기울기 (i) : -60.0003532839 degrees
근일점 방향각 (φ₀) : 4.5199852469 degrees
궤도상의 각도 (φ) : 40.4662526670 degrees
궤도 이심률 (e) : 0.4699954451
쌍성 총질량 (M) : 0.3643278901 Solar Mass


.



.

번외 1)

In [ ]:
 # 첫 번째 방정식 : 문제에서 주어진 $\frac{v_{z}}{v_{y}} = - cos{i} \, \frac{cos{\phi}\, + \, ecos{\phi_{0}}} {sin{\phi} \, + \,esin{\phi_{0}}}$ 사용
    # fsolve 기능을 사용하기 위해, 이 식을 모두 좌변으로 넘기고 이를 g라는 함수라고 했을 때, g = 0으로 만드는 작업을 한다.
    # $\frac{v_y}{v_x}\, + \, cos{i} \, \frac{cos{\phi}\, + \, ecos{\phi_{0}}} {sin{\phi} \, + \,esin{\phi_{0}}} = 0$
    astrophysical_equation_1 =  (v_y / v_x) + np.cos(i_radian) * (np.cos(phi_radian) + e * np.cos(phi0_radian)) / (np.sin(phi_radian) + e * np.sin(phi0_radian))

처음에는 코드를 짰었다. 그냥 문제에서 주어진 식, $\frac{v_{z}}{v_{y}} = - cos{i} \, \frac{cos{\phi}\, + \, ecos{\phi_{0}}} {sin{\phi} \, + \,esin{\phi_{0}}}$을 단순히 이항한 다음에, $\frac{v_y}{v_x}\, + \, cos{i} \, \frac{cos{\phi}\, + \, ecos{\phi_{0}}} {sin{\phi} \, + \,esin{\phi_{0}}} = 0$ 이것을 astrophysical_equation_1로 두었다. 근데 이제 제출하려고 vscode에서 짠 코드를 collab으로 옮기는데, 결과가 아래와 같이 나왔다.

궤도 기울기 (i) : -60.0003532839 degrees

근일점 방향각 (φ₀) : -128.4448048878 degrees

궤도상의 각도 (φ) : 40.4662526670 degrees

궤도 이심률 (e) : 0.3319208489

쌍성 총질량 (M) : 0.7459142029 Solar Mass

/tmp/ipython-input-15-940462298.py:112: Runtime Warning: The iteration is not making good progress, as measured by the improvement from the last ten iterations.result = fsolve(astrophysical_equation, initial_value)

오류의 원인은 세 번째 방정식인 Vis-viva 방정식인 (v^2 = GM(2/r - 1/a)를 사용할 때, astrophysical_equation_3 = v_magnitude_square * p - G * M_kg * (2 * p / r_magnitude - (1 - e**2))에서, 'v_magnitude_square'로 이미 v를 제곱한 값을 써놓고, 여기에도 제곱을 해서 v^4이 되니까 오류가 났던 것으로 판단된다. 왜냐하면 이 부분을 고치니까 오류 메시지도 안 뜨고, 결과도 잘 나왔기 때문이다. 그래서 처음에는 엄청나게 당황했지만, 이런 오류 덕분에 코드를 다시금 되짚어볼 수 있었다. 그러면서, 번외로 이야기하는 첫 번째 방정식의 구성에 대해 살펴보게 되었고, 이를 최종 코드에서는 변경했다.

왜냐하면 두 번째 방정식의 각주에서 말한 '그 이유는, 이제 뒤에서 할 fsolve는 반복적으로 다양한 시도값을 넣으며 해를 찾는데, 이 과정에서 만약에 M이 0에 가까운 값으로 잠깐이라도 접근하면, 원래 형태의 Vis-viva 방정식 v^2 = GM(2/r - 1/a)의 우변이 너무 작아져서 수치적으로 불안정해진다.'와 같이, 수치적 불안정의 가능성을 덜기 위해서였다.

지금 fsolve를 통해서, e와 \phi0값이 얻어지는데,  $\frac{v_y}{v_x}\, + \, cos{i} \, \frac{cos{\phi}\, + \, ecos{\phi_{0}}} {sin{\phi} \, + \,esin{\phi_{0}}} = 0$ 이 식을 사용하면, e와 $\phi_{0}$가 들어 있는 $sin{\phi}\, + \, esin{\phi_{0}}$ 값이 0에 가까워지면 분수 부분이 매우 커져서 수치적으로 불안정해질 수 있을 수도 있기 때문에, 오류를 발견하기 전에는 이 부분을 수정했었다. 오류는 다른 부분에서 발견했고, 이 부분을 바꾸지 않아도 결과는 달라지지 않았지만, 그래도 이왕 짜본 것 이전의 것은 번외로 남기고, 수정한 것을 최종으로 제출하는 바이다.